trin-core/src/types/accumulator.rs

```diff
@@ -0,0 +1,214 @@
+        // get the previous total difficulty
+        let last_total_difficulty = match self.current_epoch.header_records.len() {
+            // genesis
+            0 => U256::from_dec_str("0").unwrap(),
+            _ => {
+                self.current_epoch
+                    .header_records
+                    .last()
+                    .unwrap()
+                    .total_difficulty
+            }
+        };
+
+        // check if the epoch accumulator is full
+        if self.current_epoch.header_records.len() == EPOCH_SIZE as usize {
+            // compute the final hash for this epoch
+            let epoch_hash = self.current_epoch.tree_hash_root();
+            // append the hash for this epoch to the list of historical epochs
+            self.historical_epochs.epochs.push(epoch_hash);
+            // initialize a new empty epoch
+            self.current_epoch = EpochAccumulator {
+                header_records: vec![],
+            };
+        }
+
+        // construct the concise record for the new header and add it to the current epoch
+        let header_record = HeaderRecord {
+            block_hash: new_block_header.hash(),
+            total_difficulty: last_total_difficulty + new_block_header.difficulty,
+        };
+        self.current_epoch.header_records.push(header_record);
+    }
+}
+
+// ssz list
+// max_length = MAX_HISTORICAL_EPOCHS
+#[derive(Clone, Debug, Decode, Encode)]
+pub struct HistoricalEpochs {
+    epochs: Vec<tree_hash::Hash256>,
+}
+
+impl TreeHash for HistoricalEpochs {
+    fn tree_hash_type() -> tree_hash::TreeHashType {
+        tree_hash::TreeHashType::List
+    }
+
+    fn tree_hash_packed_encoding(&self) -> Vec<u8> {
+        unreachable!("List should never be packed.")
+    }
+
+    fn tree_hash_packing_factor() -> usize {
+        unreachable!("List should never be packed.")
+    }
+
+    fn tree_hash_root(&self) -> tree_hash::Hash256 {
+        // list of composite objects
+        // mix_in_length(merkleize([hash_tree_root(element) for element in value], limit=chunk_count(type)), len(value))
+        let hash_tree_roots: Vec<tree_hash::Hash256> = self
+            .clone()
+            .epochs
+            .into_iter()
+            .map(|epoch_root| epoch_root.tree_hash_root())
+            .collect();
+        let mut historical_epochs_hasher =
+            MerkleHasher::with_leaves(MAX_HISTORICAL_EPOCHS as usize);
+        for root in &hash_tree_roots {
+            let hash = tree_hash::merkle_root(root.as_bytes(), 0);
+            historical_epochs_hasher.write(hash.as_bytes()).unwrap();
+        }
+        let historical_epochs_root = historical_epochs_hasher.finish().unwrap();
+        tree_hash::mix_in_length(&historical_epochs_root, hash_tree_roots.len())
+    }
+}
+
+// ssz list
+// max_length = EPOCH_SIZE
+#[derive(Clone, Debug, Decode, Encode)]
+pub struct EpochAccumulator {
+    header_records: Vec<HeaderRecord>,
+}
```

Comment:
Same as above. Can we use [`VariableList`](https://docs.rs/eth2_ssz_types/0.2.1/ssz_types/struct.VariableList.html) to enforce max length?